homeassistant/components/nws/__init__.py

```diff
@@ -1 +1,186 @@
+            hass.async_create_task(
+                discovery.async_load_platform(hass, component, DOMAIN, {}, config)
+            )
+
+    return True
+
+
+class NwsData:
+    """Data class for National Weather Service integration."""
+
+    def __init__(self, hass, latitude, longitude, api_key, websession):
+        """Initialize the data."""
+        self.hass = hass
+        self.latitude = latitude
+        self.longitude = longitude
+        ha_api_key = f"{api_key} homeassistant"
+        self.nws = SimpleNWS(latitude, longitude, ha_api_key, websession)
+
+        self.update_observation_success = True
+        self.update_forecast_success = True
+        self.update_forecast_hourly_success = True
+
+    async def async_set_station(self, station):
+        """
+        Set to desired station.
+
+        If None, nearest station is used.
+        """
+        await self.nws.set_station(station)
+        _LOGGER.debug("Nearby station list: %s", self.nws.stations)
+
+    @property
+    def station(self):
+        """Return station name."""
+        return self.nws.station
+
+    @property
+    def observation(self):
+        """Return observation."""
+        return self.nws.observation
+
+    @property
+    def forecast(self):
+        """Return day+night forecast."""
+        return self.nws.forecast
+
+    @property
+    def forecast_hourly(self):
+        """Return hourly forecast."""
+        return self.nws.forecast_hourly
+
+    @staticmethod
+    async def _async_update_item(update_call, update_type, station_name, success):
+        try:
+            _LOGGER.debug("Updating %s for station %s", update_type, station_name)
+            await update_call()
+
+            if success:
+                _LOGGER.warning(
+                    "Success updating %s for station %s", update_type, station_name
+                )
+                success = True
+        except (aiohttp.ClientError, asyncio.TimeoutError) as err:
+            if success:
+                _LOGGER.warning(
+                    "Error updating %s for station %s: %s",
+                    update_type,
+                    station_name,
+                    err,
+                )
+            success = False
+
+    async def async_update(self, now=None):
+        """Update all data."""
+
+        await self._async_update_item(
+            self.nws.update_observation,
+            "observation",
+            self.station,
+            self.update_observation_success,
+        )
+        await self._async_update_item(
+            self.nws.update_forecast,
+            "forecast",
+            self.station,
+            self.update_forecast_success,
+        )
+        await self._async_update_item(
+            self.nws.update_forecast_hourly,
+            "forecast_hourly",
+            self.station,
+            self.update_forecast_hourly_success,
+        )
+
+        async_dispatcher_send(self.hass, unique_id(self.latitude, self.longitude))
```

Comment:
It looks like you are using the unique id here as well. I'm not sure dispatcher is unique to the domain so you may need to keep it here.

@balloob will know for sure.

Reply:
dispatcher is global so should keep using `DOMAIN`.